src/Avalonia.Base/ValueStore.cs

```diff
@@ -1,205 +1,231 @@
-                        return;
-                    }
-                    else
+                    if (v.Value.HasValue)
                     {
-                        priorityValue = CreatePriorityValue(property);
-                        priorityValue.SetValue(v, (int)BindingPriority.LocalValue);
-                        _propertyValues.SetValue(property, priorityValue);
+                        value = v.Value.Value;
+                        return true;
                     }
                 }
-            }
-            else
-            {
-                if (value == AvaloniaProperty.UnsetValue)
+                else
                 {
-                    return;
+                    value = (T)slot;
+                    return true;
                 }
+            }
+
+            value = default!;
+            return false;
+        }
 
-                if (priority == (int)BindingPriority.LocalValue)
+        public bool TryGetValueUntyped(AvaloniaProperty property, out object? value)
+        {
+            if (_values.TryGetValue(property, out var slot))
+            {
+                if (slot is IValue v)
                 {
-                    Validate(property, ref value);
-                    _propertyValues.AddValue(property, value);
-                    Changed(property, priority, AvaloniaProperty.UnsetValue, value);
-                    return;
+                    if (v.Value.HasValue)
+                    {
+                        value = v.Value.Value;
+                        return true;
+                    }
                 }
                 else
                 {
-                    priorityValue = CreatePriorityValue(property);
-                    _propertyValues.AddValue(property, priorityValue);
+                    value = slot;
+                    return true;
                 }
             }
 
-            priorityValue.SetValue(value, priority);
-        }
-
-        public void BindingNotificationReceived(AvaloniaProperty property, BindingNotification notification)
-        {
-            _owner.BindingNotificationReceived(property, notification);
-        }
-
-        public void Changed(AvaloniaProperty property, int priority, object oldValue, object newValue)
-        {
-            _owner.PriorityValueChanged(property, priority, oldValue, newValue);
+            value = default;
+            return false;
         }
 
-        public IDictionary<AvaloniaProperty, object> GetSetValues()
+        public void SetValue<T>(StyledPropertyBase<T> property, T value, BindingPriority priority)
         {
-            return _propertyValues.ToDictionary();
+            if (_values.TryGetValue(property, out var slot))
+            {
+                SetExisting(slot, property, value, priority);
+            }
+            else if (priority == BindingPriority.LocalValue)
+            {
+                _values.AddValue(property, (object)value!);
```

Comment:
This is causing boxing of all non-reference types values that have local priority. Is local value priority not important from performance perspective?

Reply:
This was actually my attempt at an optimization, ported from the old `ValueStore`. The idea being that the most common case is a single local value: rather than creating a new `ConstantValueEntry` to wrap the value, just box the value and store it directly in the `AvaloniaPropertyValueStore`. In the absence discriminated unions in C#, `AvaloniaPropertyValueStore`'s value type must be heap allocated anyway.

I'd be interested in seeing the quick hack for storing typed local values you mentioned.

Reply:
Sure, when I come back from work I can share it. Hopefully I still have it around :) But I think the gist of it was implementing `IValue` and reusing that instead of boxing.

Reply:
Delivering hack payload: https://github.com/MarchingCube/Avalonia/commit/08318164a73e64661eb177b011981ff20ba50dc8.

Maybe I missed something but it allocates less and runs faster on my PC.

Reply:
Yep, you're right! I implemented something based on your hack and `Set_Int_Property_LocalValue`  allocated memory goes from 2632B -> 256B and faster to boot. Nice ;)